Client/game_sa/CDummyPoolSA.cpp

```diff
@@ -13,64 +13,96 @@
 {
-    if (m_pLodBackup)
+    if (m_pOriginalElementsBackup)
         return;
 
-    m_pLodBackup = std::make_unique<std::array<CEntitySAInterface*, MAX_DUMMIES>>();
+    m_pOriginalElementsBackup = std::make_unique<pool_backup_t>();
 
-    for (int i = 0; i < MAX_DUMMIES; i++)
+    auto pDummyPool = (*m_ppDummyPoolInterface);
+    for (auto i = 0; i < MAX_DUMMIES_DEFAULT; i++)
     {
-        CEntitySAInterface* object = (*m_ppDummyPoolInterface)->GetObject(i);
-        (*m_pLodBackup)[i] = object->m_pLod;
-        object->m_pLod = nullptr;
+        if (pDummyPool->IsContains(i))
+        {
+            CEntitySAInterface* building = pDummyPool->GetObject(i);
+
+            pGame->GetWorld()->Remove(building, CDummyPool_Destructor);
+            building->RemoveRWObjectWithReferencesCleanup();
+
+            pDummyPool->Release(i);
+
+            (*m_pOriginalElementsBackup)[i].first = true;
+            (*m_pOriginalElementsBackup)[i].second = *building;
+        }
+        else
+        {
+            (*m_pOriginalElementsBackup)[i].first = false;
+        }
     }
 }
 
-void CDummyPoolSA::RestoreAllBuildingsLods()
+void CDummyPoolSA::RestoreBackup()
 {
-    if (!m_pLodBackup)
+    if (!m_pOriginalElementsBackup)
         return;
 
-    for (int i = 0; i < MAX_DUMMIES; i++)
+    auto& originalData = *m_pOriginalElementsBackup;
+    auto  pDummyPool = (*m_ppDummyPoolInterface);
+    for (auto i = 0; i < MAX_DUMMIES_DEFAULT; i++)
     {
-        CEntitySAInterface* object = (*m_ppDummyPoolInterface)->GetObject(i);
-        object->m_pLod = (*m_pLodBackup)[i];
+        if (originalData[i].first)
+        {
+            pDummyPool->AllocateAt(i);
+            auto pDummy = pDummyPool->GetObject(i);
+            *pDummy = originalData[i].second;
+
+            pGame->GetWorld()->Add(pDummy, CDummyPool_Constructor);
+        }
     }
 
-    m_pLodBackup.release();
+    m_pOriginalElementsBackup = nullptr;
 }
 
-void CDummyPoolSA::UpdateBuildingLods(void* oldPool, void* newPool)
+void CDummyPoolSA::UpdateBuildingLods(const std::uint32_t offset)
 {
-    const uint32_t offset = (uint32_t)newPool - (uint32_t)oldPool;
+    if (m_pOriginalElementsBackup)
+        UpdateBackupLodOffset(offset);
+    else
+        UpdateLodsOffestInPool(offset);
+}
 
-    if (m_pLodBackup)
+void CDummyPoolSA::UpdateBackupLodOffset(const std::uint32_t offset)
+{
+    for (auto i = 0; i < (*m_pOriginalElementsBackup).size(); i++)
     {
-        for (int i = 0; i < MAX_DUMMIES; i++)
+        if ((*m_pOriginalElementsBackup)[i].first)
         {
-            if ((*m_pLodBackup)[i] != nullptr)
-            {
-                (*m_pLodBackup)[i] = (CEntitySAInterface*)((uint32_t)(*m_pLodBackup)[i] + offset);
-            }
+            CEntitySAInterface* object = &(*m_pOriginalElementsBackup)[i].second;
+            CEntitySAInterface* lod = object->GetLod();
+            if (lod)
+                object->SetLod((CEntitySAInterface*)((std::uint32_t)lod + offset));
```

Comment:
Use range loop instead, also use early continue (early return)

Reply:
Addressed in 71d99cb